File: client/src/main/java/net/hoz/netapi/client/service/NetGameProvider.java
package net.hoz.netapi.client.service;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.google.inject.Inject;
import com.google.protobuf.StringValue;
import com.iamceph.resulter.core.DataResultable;
import com.iamceph.resulter.core.Resultable;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import net.hoz.api.Controlled;
import net.hoz.api.data.ResultableData;
import net.hoz.api.data.WUUID;
import net.hoz.api.data.game.GameConfig;
import net.hoz.api.data.game.ProtoGameFrame;
import net.hoz.api.data.game.ProtoSpawnerType;
import net.hoz.api.data.game.StoreHolder;
import net.hoz.api.service.MGameType;
import net.hoz.api.service.NetGameServiceClient;
import net.hoz.api.util.Packeto;
import net.hoz.api.util.ReactorHelper;
import net.hoz.netapi.client.config.DataConfig;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Collection;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Provider for {@link ProtoGameFrame} and other game-related data.
 * <p>
 * This provider communicates with our BAGR backend heavily and when the provider is initialized,
 * it asks BAGR about all data related to given {@link net.hoz.api.data.GameType}.
 */
@Slf4j
public class NetGameProvider implements Controlled {
    /**
     * Stores name-to-uuid values for games.
     */
    private final Map<String, UUID> gameNameToUUID = new ConcurrentHashMap<>();
    /**
     * Cached games.
     */
    @Getter
    private final Cache<UUID, ProtoGameFrame> gameCache = Caffeine.newBuilder()
            .build();
    /**
     * Cached game configurations.
     */
    @Getter
    private final Cache<String, GameConfig> configCache = Caffeine.newBuilder()
            .build();
    /**
     * Cached game stores.
     */
    @Getter
    private final Cache<String, StoreHolder> storeCache = Caffeine.newBuilder()
            .build();
    /**
     * Cached game spawner types.
     */
    @Getter
    private final Cache<String, ProtoSpawnerType> spawnerCache = Caffeine.newBuilder()
            .build();
    /**
     * RSocket service for communicating with BAGR.
     */
    private final NetGameServiceClient gameService;
    /**
     * A {@link net.hoz.api.data.GameType} message for protobuf.
     */
    private final MGameType gameTypeMessage;

    /**
     * Main constructor.
     *
     * @param gameService  RSocket game client.
     * @param clientConfig simple configuration
     */
    @Inject
    public NetGameProvider(NetGameServiceClient gameService,
                           DataConfig clientConfig) {
        this.gameService = gameService;
        this.gameTypeMessage = MGameType.newBuilder()
                .setType(clientConfig.gameType())
                .build();
    }

    @Override
    public void initialize() {
        subscribeForUpdates();
        createDataCache();
    }

    @Override
    public void dispose() {
        //TODO: handle destroying too
    }

    /**
     * Tries to get one game with given ID from cache.
     *
     * @param gameId id of the game
     * @return {@link DataResultable} with result
     */
    public DataResultable<ProtoGameFrame> oneGame(UUID gameId) {
        return DataResultable.failIfNull(gameCache.getIfPresent(gameId), "Game not found.");
    }

    /**
     * Tries to get one game with given name from cache.
     *
     * @param name name of the game
     * @return {@link DataResultable} with result
     */
    public DataResultable<ProtoGameFrame> oneGame(String name) {
        final var uuid = gameNameToUUID.get(name);
        if (uuid == null) {
            return DataResultable.fail("Game not found.");
        }
        return oneGame(uuid);
    }

    /**
     * Gets all available games from cache.
     *
     * @return Collection of games.
     */
    public Collection<ProtoGameFrame> allGames() {
        return gameCache.asMap().values();
    }

    /**
     * Tries to get one config with given name from cache.
     *
     * @param name name of the config
     * @return {@link DataResultable} with result
     */
    public DataResultable<GameConfig> oneConfig(String name) {
        return DataResultable.failIfNull(configCache.getIfPresent(name), "Config not found.");
    }

    /**
     * Gets all available configs from cache.
     *
     * @return Collection of configs.
     */
    public Collection<GameConfig> allConfigs() {
        return configCache.asMap().values();
    }

    /**
     * Tries to get one store with given name from cache.
     *
     * @param name name of the store
     * @return {@link DataResultable} with result
     */
    public DataResultable<StoreHolder> oneStore(String name) {
        return DataResultable.failIfNull(storeCache.getIfPresent(name), "Store not found.");
    }

    /**
     * Gets all available stores from cache.
     *
     * @return Collection of stores.
     */
    public Collection<StoreHolder> allStores() {
        return storeCache.asMap().values();
    }

    /**
     * Tries to get one spawner with given name from cache.
     *
     * @param name name of the spawner
     * @return {@link DataResultable} with result
     */
    public DataResultable<ProtoSpawnerType> oneSpawner(String name) {
        return DataResultable.failIfNull(spawnerCache.getIfPresent(name), "Spawner not found.");
    }

    /**
     * Gets all available spawners from cache.
     *
     * @return Collection of spawners.
     */
    public Collection<ProtoSpawnerType> allSpawners() {
        return spawnerCache.asMap().values();
    }

    /**
     * Tries to retrieve the game from BAGR backend.
     *
     * @param gameId game id
     * @return {@link DataResultable} of the operation.
     */
    public Mono<DataResultable<ProtoGameFrame>> loadGame(UUID gameId) {
        return doGameLoading(gameService.oneById(
                WUUID.newBuilder()
                        .setValue(gameId.toString())
                        .build())
        );
    }

    /**
     * Tries to retrieve the game from backend.
     *
     * @param name name of the game
     * @return {@link DataResultable} of the operation.
     */
    public Mono<DataResultable<ProtoGameFrame>> loadGame(String name) {
        return doGameLoading(gameService.oneByName(
                StringValue.newBuilder()
                        .setValue(name)
                        .build())
        );
    }

    /**
     * Tries to save given game to the backend and replaces it in the cache.
     *
     * @param frame game to save.
     * @return {@link DataResultable} result of this operation.
     */
    public Mono<DataResultable<UUID>> saveGame(ProtoGameFrame frame) {
        //TODO: check this
        return gameService.saveGame(frame)
                .map(next -> DataResultable.from(next.getResult(), UUID.fromString(frame.getUuid())))
                .doOnNext(next -> next.ifOk(data -> gameCache.put(data, frame)))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Tries to save given spawner to the backend and caches it.
     *
     * @param spawnerTypeHolder holder to save
     * @return {@link Resultable} result of this operation.
     */
    public Mono<Resultable> saveSpawnerType(ProtoSpawnerType spawnerTypeHolder) {
        final var spawnerName = spawnerTypeHolder.getName();
        return gameService.saveSpawnerType(spawnerTypeHolder)
                .map(Resultable::convert)
                .doOnNext(next -> {
                    if (next.isOk()) {
                        spawnerCache.put(spawnerName, spawnerTypeHolder);
                        log.debug("Saved new spawner[{}] for GameType[{}].", spawnerName, gameTypeMessage.getType());
                    }
                })
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Retrieves all available games from backend.
     *
     * @return Flux of {@link ProtoGameFrame}
     */
    public Flux<ProtoGameFrame> loadGames() {
        return gameService.all(gameTypeMessage)
                .doOnNext(next -> log.trace("Received game [{}] - [{}] for GameType[{}].", next.getName(), next.getUuid(), next.getType()))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Retrieves all available configs from backend.
     *
     * @return Flux of {@link GameConfig}
     */
    public Flux<GameConfig> loadConfigs() {
        return gameService.allConfigs(gameTypeMessage)
                .doOnNext(next -> log.trace("Received config[{}] for GameType[{}].", next.getName(), next.getType()))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Retrieves all available stores from backend.
     *
     * @return Flux of {@link StoreHolder}
     */
    public Flux<StoreHolder> loadStores() {
        return gameService.allStores(gameTypeMessage)
                .doOnNext(next -> log.trace("Received store[{}] for GameType[{}].", next.getName(), next.getGameType()))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Retrieves all available spawners from backend.
     *
     * @return Flux of {@link ProtoSpawnerType}
     */
    public Flux<ProtoSpawnerType> loadSpawners() {
        return gameService.allSpawnerTypes(gameTypeMessage)
                .doOnNext(next -> log.trace("Received store[{}] for GameType[{}].", next.getName(), next.getType()))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Processes the given {@link ResultableData} into a {@link ProtoGameFrame}.
     *
     * @param loadingMono mono from the backend
     * @return mono with {@link DataResultable} result of the operation.
     */
    protected Mono<DataResultable<ProtoGameFrame>> doGameLoading(Mono<ResultableData> loadingMono) {
        return loadingMono
                .filter(ReactorHelper.filterResult(log))
                .map(result -> Packeto.unpack(result.getData(), ProtoGameFrame.class))
                .doOnNext(next -> next.ifOk(game -> {
                    final var uuid = UUID.fromString(game.getUuid());
                    log.debug("Received game [{}] - [{}] for GameType[{}]", game.getName(), uuid, game.getType());
                    gameCache.put(uuid, game);
                }))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Loads all cacheable values from the backend.
     * The order shouldn't be important at all.
     */
    protected void createDataCache() {
        loadGames()
                .doOnNext(next -> {
                    log.trace("Caching new game[{}]...", next.getName());

                    final var gameId = UUID.fromString(next.getUuid());
                    gameCache.put(gameId, next);
                    gameNameToUUID.put(next.getName(), gameId);
                })
                .onErrorResume(ex -> ReactorHelper.fluxError(ex, log))
                .subscribe();

        loadConfigs()
                .doOnNext(next -> {
                    log.trace("Caching new config[{}]...", next.getName());
                    configCache.put(next.getName(), next);
                })
                .onErrorResume(ex -> ReactorHelper.fluxError(ex, log))
                .subscribe();

        loadStores()
                .doOnNext(next -> {
                    log.trace("Caching new store[{}]...", next.getName());
                    storeCache.put(next.getName(), next);
                })
                .onErrorResume(ex -> ReactorHelper.fluxError(ex, log))
                .subscribe();

        loadSpawners()
                .doOnNext(next -> {
                    log.trace("Caching new spawner type[{}]...", next.getName());
                    spawnerCache.put(next.getName(), next);
                })
                .onErrorResume(ex -> ReactorHelper.fluxError(ex, log))
                .subscribe();
    }

    protected void subscribeForUpdates() {

    }
}

File: client/src/main/kotlin/net/hoz/netapi/client/provider/NetGameProvider.kt
/*
 * Copyright 2022 hoz-network.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.hoz.netapi.client.provider

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import com.iamceph.resulter.core.DataResultable
import com.iamceph.resulter.core.Resultable
import com.iamceph.resulter.core.model.ResultableData
import com.iamceph.resulter.kotlin.resultable
import com.iamceph.resulter.kotlin.unpack
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import mu.KotlinLogging
import net.hoz.api.data.GameType
import net.hoz.api.data.WUUID
import net.hoz.api.data.game.GameConfig
import net.hoz.api.data.game.ProtoGameFrame
import net.hoz.api.data.game.ProtoSpawnerType
import net.hoz.api.data.game.StoreHolder
import net.hoz.api.data.wUUID
import net.hoz.api.service.NetGameServiceGrpcKt
import net.hoz.api.service.gameByNameRequest
import net.hoz.api.service.mGameType
import net.hoz.netapi.api.Controlled
import net.hoz.netapi.api.onErrorHandle
import net.hoz.netapi.client.config.DataConfig
import net.hoz.netproto.asUuid
import network.hoz.kaffeine.get
import network.hoz.kaffeine.set
import java.util.*
import javax.inject.Inject
import kotlin.coroutines.coroutineContext

private val log = KotlinLogging.logger {}

class NetGameProvider @Inject constructor(
    private val gameService: NetGameServiceGrpcKt.NetGameServiceCoroutineStub,
    private val clientConfig: DataConfig
) : Controlled {

    /**
     * Stores name-to-uuid values for games.
     */
    private val gameNameToUUID: MutableMap<String, UUID> = mutableMapOf()
    private val gameCache: Cache<UUID, ProtoGameFrame> = Caffeine.newBuilder().build()
    private val configCache: Cache<String, GameConfig> = Caffeine.newBuilder().build()
    private val storeCache: Cache<String, StoreHolder> = Caffeine.newBuilder().build()
    private val spawnerCache: Cache<String, ProtoSpawnerType> = Caffeine.newBuilder().build()

    private var gameUpdateJob: Job? = null

    override suspend fun initialize() {
        gameUpdateJob?.cancel("Staring new listening for Game updates")

        gameUpdateJob = coroutineScope {
            launch { subscribeForUpdates() }
        }
    }

    override fun dispose() {
        gameUpdateJob?.cancel("Ending listening for Game updates.")
        TODO("Not yet implemented")
    }

    /**
     * Tries to get one game with given ID from cache.
     *
     * @param gameId id of the game
     * @return [DataResultable] with result
     */
    fun oneGame(gameId: UUID): DataResultable<ProtoGameFrame> =
        DataResultable.failIfNull(gameCache[gameId], "Game not found.")

    /**
     * Tries to get one game with given name from cache.
     *
     * @param name name of the game
     * @return [DataResultable] with result
     */
    fun oneGame(name: String): DataResultable<ProtoGameFrame> =
        gameNameToUUID[name]?.let { oneGame(it) } ?: DataResultable.fail("Game not found.")

    /**
     * Gets all available games from cache.
     *
     * @return Collection of games.
     */
    fun allGames(): Collection<ProtoGameFrame> = gameCache.asMap().values // TODO: add extension for Cache#values

    /**
     * Tries to get one config with given name from cache.
     *
     * @param name name of the config
     * @return [DataResultable] with result
     */
    fun oneConfig(name: String): DataResultable<GameConfig> =
        DataResultable.failIfNull(configCache[name], "Config not found.")

    /**
     * Gets all available configs from cache.
     *
     * @return Collection of configs.
     */
    fun allConfigs(): Collection<GameConfig> = configCache.asMap().values // TODO: add extension for Cache#values

    /**
     * Tries to get one store with given name from cache.
     *
     * @param name name of the store
     * @return [DataResultable] with result
     */
    fun oneStore(name: String): DataResultable<StoreHolder> =
        DataResultable.failIfNull(storeCache[name], "Store not found.")

    /**
     * Gets all available stores from cache.
     *
     * @return Collection of stores.
     */
    fun allStores(): Collection<StoreHolder> = storeCache.asMap().values // TODO: add extension for Cache#values

    /**
     * Tries to get one spawner with given name from cache.
     *
     * @param name name of the spawner
     * @return [DataResultable] with result
     */
    fun oneSpawner(name: String): DataResultable<ProtoSpawnerType> =
        DataResultable.failIfNull(spawnerCache[name], "Spawner not found.")

    /**
     * Gets all available spawners from cache.
     *
     * @return Collection of spawners.
     */
    fun allSpawners(): Collection<ProtoSpawnerType> =
        spawnerCache.asMap().values // TODO: add extension for Cache#values

    /**
     * Tries to retrieve the game from BAGR backend.
     *
     * @param gameId game id
     * @return [DataResultable] of the operation.
     */
    suspend fun loadGame(gameId: UUID): DataResultable<ProtoGameFrame> =
        handleGameData(
            gameService.oneById(
                wUUID { gameId.toString() }
            )
        )

    /**
     * Tries to retrieve the game from backend.
     *
     * @param name name of the game
     * @return [DataResultable] of the operation.
     */
    suspend fun loadGame(name: String): DataResultable<ProtoGameFrame> =
        handleGameData(
            gameService.oneByName(
                gameByNameRequest {
                    this.name = name
                    this.type = clientConfig.gameType
                }
            )
        )

    /**
     * Tries to save given game to the backend and replaces it in the cache.
     *
     * @param frame game to save.
     * @return [DataResultable] result of this operation.
     */
    // TODO: check this
    suspend fun saveGame(frame: ProtoGameFrame): DataResultable<UUID> {
        return gameService.saveGame(frame)
            .unpack(WUUID::class)
            .map { it.asUuid() }
            .ifOk { gameCache[it] = frame }
    }

    /**
     * Tries to save given spawner to the backend and caches it.
     *
     * @param spawnerTypeHolder holder to save
     * @return [Resultable] result of this operation.
     */
    suspend fun saveSpawnerType(spawnerTypeHolder: ProtoSpawnerType): Resultable {
        return gameService.saveSpawnerType(spawnerTypeHolder)
            .resultable()
            .also {
                if (it.isOk) {
                    val spawnerName = spawnerTypeHolder.name

                    spawnerCache[spawnerName] = spawnerTypeHolder
                    log.debug { "Saved new spawner [$spawnerName] for GameType[${spawnerTypeHolder.type}]." }
                }
            }
    }

    /**
     * Retrieves all available games from backend.
     *
     * @return Flux of [ProtoGameFrame]
     */
    fun loadGames(type: GameType): Flow<ProtoGameFrame> = gameService.all(mGameType { this.type = type })
        .onEach { log.debug { "Received game [${it.name}] - [${it.uuid}] for GameType[${it.type}]." } }
        .onErrorHandle(log)


    /**
     * Retrieves all available configs from backend.
     *
     * @return Flux of [GameConfig]
     */
    fun loadConfigs(type: GameType): Flow<GameConfig> = gameService.allConfigs(mGameType { this.type = type })
        .onEach { log.debug { "Received config [${it.name}] for GameType[${it.type}]." } }
        .onErrorHandle(log)

    /**
     * Retrieves all available stores from backend.
     *
     * @return Flux of [StoreHolder]
     */
    fun loadStores(type: GameType): Flow<StoreHolder> = gameService.allStores(mGameType { this.type = type })
        .onEach { log.debug { "Received store holder [${it.name}] - [${it.id}] for GameType[${it.gameType}]." } }
        .onErrorHandle(log)

    /**
     * Retrieves all available spawners from backend.
     *
     * @return Flux of [ProtoSpawnerType]
     */
    fun loadSpawners(type: GameType): Flow<ProtoSpawnerType> =
        gameService.allSpawnerTypes(mGameType { this.type = type })
            .onEach { log.debug { "Received spawner [${it.name}] for GameType[${it.type}]." } }
            .onErrorHandle(log)

    /**
     * Processes the given [ResultableData] into a [ProtoGameFrame].
     *
     * @param loadingMono mono from the backend
     * @return mono with [DataResultable] result of the operation.
     */
    private fun handleGameData(data: ResultableData): DataResultable<ProtoGameFrame> =
        data.unpack(ProtoGameFrame::class)
            .ifOk {
                val uuid = UUID.fromString(it.uuid)

                log.debug { "Received game [${it.name}] - [$uuid] for GameType[${it.type}]" }
                gameCache[uuid] = it
            }

    /**
     * Loads all cacheable values from the backend.
     * The order shouldn't be important at all.
     */
    suspend fun fetchData(type: GameType) {
        loadGames(type)
            .onEach {
                val gameId = UUID.fromString(it.uuid)

                log.trace { "Caching new game [${it.name}]..." }
                gameCache[gameId] = it
                gameNameToUUID[it.name] = gameId
            }
            .collect()

        loadConfigs(type)
            .onEach {
                log.trace { "Caching new config [${it.name}]..." }
                configCache[it.name] = it
            }
            .onErrorHandle(log)
            .collect()

        loadStores(type)
            .onEach {
                log.trace { "Caching new store [${it.name}]..." }
                storeCache[it.name] = it
            }
            .onErrorHandle(log)
            .collect()

        loadSpawners(type)
            .onEach {
                log.trace { "Caching new spawner type [${it.name}]..." }
                spawnerCache[it.name] = it
            }
            .onErrorHandle(log)
            .collect()
    }

    private suspend fun subscribeForUpdates() {
        //TODO: handle game updates
    }
}